lsat/store.go

```diff
@@ -1,38 +1,57 @@
 package lsat
 
 import (
-	"fmt"
+	"errors"
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"strings"
+	"sync"
+	"time"
 )
 
 var (
 	// ErrNoToken is the error returned when the store doesn't contain a
 	// token yet.
-	ErrNoToken = fmt.Errorf("no token in store")
+	ErrNoToken = errors.New("no token in store")
 
+	// storeFileName is the name of the file where we store the final,
+	//valid, token to.
 	storeFileName = "lsat.token"
+
+	// storeFileNamePending is the name of the file where we store a pending
+	// token until it was successfully paid for.
+	storeFileNamePending = "lsat.token.pending"
+
+	// errNoReplace is the error that is returned if a pending token is
+	// being written to a store that already contains a paid token.
+	errNoReplace = errors.New("won't replace existing token with " +
+		"pending token. " + manualRetryHint)
 )
 
 // Store is an interface that allows users to store and retrieve an LSAT token.
 type Store interface {
-	// HasToken returns true if the store contains a token.
-	HasToken() bool
+	// CurrentToken returns the token that is currently contained in the
+	// store or an error if there is none.
+	CurrentToken() (*Token, error)
 
-	// Token returns the token that is contained in the store or an error
-	// if there is none.
-	Token() (*Token, error)
+	// AllTokens returns all tokens that the store has knowledge of, even
+	// if they might be expired. The tokens are mapped by their identifying
+	// attribute like file name or storage key.
+	AllTokens() (map[string]*Token, error)
 
-	// StoreToken saves a token to the store, overwriting any old token if
-	// there is one.
+	// StoreToken saves a token to the store. Old tokens should be kept for
+	// accounting purposes but marked as invalid somehow.
 	StoreToken(*Token) error
 }
 
-// FileStore is an implementation of the Store interface that uses a single file
-// to save the serialized token.
+// FileStore is an implementation of the Store interface that files to save the
+// serialized tokens. There is always just one current token that is either
+// pending or fully paid.
 type FileStore struct {
-	fileName string
+	fileName        string
+	fileNamePending string
+	lock            sync.Mutex
```

Comment:
Yeah, you're right. I added it because @Roasbeef suggested to do so. But as long as we don't use the store outside of the interceptor it doesn't add any value.